t/sendreport/multiple-send-methods.t

```diff
@@ -0,0 +1,516 @@
+    is $report->external_id, 248, 'Report has external ID';
+
+    is $report->send_fail_count, 4, 'send_fail_count unchanged';
+    is $report->send_fail_reason, 'Open311 fail',
+        'send_fail_reason unchanged';
+    cmp_bag $report->send_fail_body_ids, [], 'No send_fail_body_ids';
+
+    $report->result_source->schema->cobrand->{c} = $mock_catalyst;
+    like $report->duration_string,
+        qr/Cherwell District Council and Oxfordshire County Council/,
+        'duration string mentions Cherwell and Oxford';
+};
+
+subtest 'Test resend' => sub {
+    $mock_open311->mock(
+        'send',
+        sub {
+            shift->error('Open311 fail');
+            return -1;
+        }
+    );
+
+    my ($report_for_resend) = $mech->create_problems_for_body(
+        1,
+        ( join ',', $body_oxf->id, $body_cherwell->id ),
+        'Test resend',
+        {   cobrand  => 'fixmystreet',
+            category => 'Other',
+            user     => $user,
+        },
+    );
+
+    # Send report; we expect failure for Open311
+    test_send();
+    $report_for_resend->discard_changes;
+
+    ok $report_for_resend->whensent, 'whensent recorded';
+    is $report_for_resend->send_method_used, 'Email',
+        'send_method_used recorded';
+
+    is $report_for_resend->send_fail_count, 1, 'send_fail_count recorded';
+    is $report_for_resend->send_fail_reason, 'Open311 fail',
+        'send_fail_reason recorded';
+    cmp_bag $report_for_resend->send_fail_body_ids, [ $body_oxf->id ],
+        'Failed body ID recorded';
+
+    # Call 'resend'; we expect certain fields to be unset
+    $report_for_resend->resend;
+    $report_for_resend->update;
+
+    is $report_for_resend->whensent,         undef, 'whensent unset';
+    is $report_for_resend->send_method_used, undef, 'send_method_used unset';
+
+    is $report_for_resend->send_fail_count, 1, 'send_fail_count unmodified';
+    is $report_for_resend->send_fail_reason, 'Open311 fail',
+        'send_fail_reason unmodified';
+    cmp_bag $report_for_resend->send_fail_body_ids, [],
+        'send_fail_body_ids unset';
+
+    $report->result_source->schema->cobrand->{c} = $mock_catalyst;
+    like $report->duration_string,
+        qr/Cherwell District Council and Oxfordshire County Council/,
+        'duration string mentions Cherwell and Oxford';
+};
+
+subtest 'Test staging send' => sub {
+    # Will send with send_reports flag set to 0, so email will be used
+    # instead of Open311
+
+    note 'Testing for report with 1 body (Open311)';
+
+    my ($report_for_staging) = $mech->create_problems_for_body(
+        1,
+        $body_oxf->id,
+        'Test staging send',
+        {   cobrand  => 'fixmystreet',
+            category => 'Other',
+            user     => $user,
+        },
+    );
+
+    # This forces email to be sent to user
+    # (see perllib/FixMyStreet/SendReport/Email.pm->send())
+    # CHECKME Is this a sensible thing to do?
+    FixMyStreet->test_mode(0);
```

Comment:
Not sure I should be doing this.

Reply:
The test suite runs with staging set and send_reports defaults to 0, so the test_mode flag is in SendReport::Email in order for most tests to be able to test sending emails. I'm not sure it's vital to test this non-test little bit of code, but I can't see any harm in having it - though if anything, this particular test would be in t/sendreport/email.t I guess, it's not testing multiple send methods at this point.